### ISSUE TITLE
Bit length, useful for ordinary numbers?

### ISSUE THREAD

**Issue by js-choi** [js-choi/proposal-bigint-bit-length#3]
I wrote this in the explainer: 

> In contrast to other operations – such as abs and sqrt from the [BigInt Math proposal](https://github.com/tc39/proposal-bigint-math) – the new bit-length API would probably be useful only on BigInts and not on ordinary numbers.

Is this really true? One potential use case for ordinary-number bit lengths might be conveniently determining whether an ordinary number integer will fit within e.g. an Int16Array.

Of course, this would be complicated by what to do with non-integer numbers (something to do with IEEE floating-point representation? truncate? throw a TypeError?).

Either way, before or even during Stage 1, it is worth explicitly considering bit lengths for ordinary numbers.

**Comment by ljharb** [comment]
If I needed it for an integer number, i'd do `BigInt(number).bitLength`. I don't think there's a sensible answer for non-integer Numbers, and it'd be very weird to have an accessor that throws on some Numbers and not others.

**Comment by js-choi** [comment]
I agree that “properties should not throw” is a good principle, that `.bitLength` would probably be best as a property, and that `BigInt(number).bitLength` probably is the best general solution. However, bit length is not necessarily a property; it could still be a static function like `Math.bitLength()` that throws or truncates on non-integer numbers. For example:

```js
Math.bitLength(255) // 1?
Math.bitLength(255.5) // ???
Math.bitLength(255n) // 1
```

Basically, I’m trying to think about how consistent we *should* be between this proposal, [proposal-popcount][] (which I would definitely use on 8-, 16-, and 32-bit integer numbers), as well as maybe the existing Math.clz32:

| Function                              | Useful for non-ints? | Useful for int numbers? | Useful for BigInts? | Non-int behavior         
| ------------------------------------- | -------------------- | ----------------------- | ------------------- | ------------------------ 
| Bit length                            | No                   | ???                     | Yes                 | ???                      
| [`Math.popcnt()`][proposal-popcount]* | No                   | Yes                     | Yes                 | Truncation?
| `Math.clz32()`                        | No                   | Yes                     | Yes                 | Truncation with ToUint32 

\* Name subject to change.

Is there a fundamental difference between bit length, popcount, and clz32 such that that we should make popcount and clz32 apply to integral ordinary numbers but not to bit length? (I’m thinking probably yes.) If so, then we should give bit length a special API that is unlike popcount or clz32.

In general, when JavaScript developers use the bits of integral ordinary numbers (e.g., from Int8Array), they already know the bit length that they are dealing with, (e.g., 8, 16, 32). If they do not know the narrowest word width that fits certain values, they can just coerce to a BigInt like you suggest (i.e., `BigInt(i).bitLength`) or use `<` with an exponent of 2 (e.g. `i < 2 ** 8`, `i < 2 ** 16`, `i < 2 ** 32`).

(It may be useful being able to determine whether a non-integer number fits in a 32- versus 64-bit IEEE floating-point representation, but I don’t really see a way to fit that into the same API as integers’ bit lengths.)

So I’m inclined to agree to keep bit lengths specific to BigInts and use `BigInt(i).bitLength` as necessary, but it’s probably still worth exploring in these early stages whether elegantly fitting in ordinary numbers is possible. Having said that, the answer is probably, “No, there’s no elegant way to fit ordinary numbers in”.

[proposal-popcount]: https://github.com/js-choi/proposal-popcount

**Comment by ljharb** [comment]
Either way, the committee has been very clear that they do not want things that throw on some numbers and not others, or else number/bigint mixtures would have worked much more intuitively.

Unless there's a reasonable answer for *every* number, then it doesn't belong there.

**Comment by Yaffle** [comment]
Some method like `frexp` from C library can be useful for number type, it returns floor(log2(value)) + 1 (If I am not mistaking), which is something similar to bitLength). Note: `Math.floor(Math.log2(Number.MAX_VALUE)) + 1` returns 1025, not 1024, because of the rounding, and perhaps slower.
The use case is reimplementation of Math.sin/Math.exp/... in JS itself :-) using algorithms from fdlibm - you need to extract exponent for this.
